File: sdk/ts/src/core/interface.ts
import { Address, Hex } from 'viem';
import Decimal from 'decimal.js';
import {
  State,
  ChannelDefinition,
  HomeChannelDataResponse,
  EscrowDepositDataResponse,
  EscrowWithdrawalDataResponse,
} from './types';
import {
  HomeChannelCreatedEvent,
  HomeChannelMigratedEvent,
  HomeChannelCheckpointedEvent,
  HomeChannelChallengedEvent,
  HomeChannelClosedEvent,
  EscrowDepositInitiatedEvent,
  EscrowDepositChallengedEvent,
  EscrowDepositFinalizedEvent,
  EscrowWithdrawalInitiatedEvent,
  EscrowWithdrawalChallengedEvent,
  EscrowWithdrawalFinalizedEvent,
} from './event';

// ============================================================================
// Listener Interface
// ============================================================================

/**
 * Listener defines the interface for listening to channel events
 */
export interface Listener {
  /**
   * Listen starts listening for events
   */
  listen(): Promise<void>;
}

// ============================================================================
// Client Interface
// ============================================================================

/**
 * Client defines the interface for interacting with the ChannelsHub smart contract
 */
export interface Client {
  // ========= Getters - IVault =========

  /**
   * Get balances for multiple accounts and tokens
   * @param accounts - Array of account addresses
   * @param tokens - Array of token addresses
   * @returns 2D array of balances [account][token]
   */
  getAccountsBalances(accounts: Address[], tokens: Address[]): Promise<Decimal[][]>;

  /**
   * Get parametric token sub balances for multiple accounts and subIds
   * @param accounts - Array of account addresses
   * @param token - Parametric token address
   * @param subIds - Array of subaccount IDs
   * @returns 2D array of balances [account][token]
   */
  getAccountsSubBalances(
    accounts: Address[],
    token: Address,
    subIds: number[],
  ): Promise<Decimal[][]>;

  // ========= Getters - Token Balance & Approval =========

  /**
   * Get on-chain token balance (ERC-20 or native) for a wallet
   * @param asset - Asset symbol (e.g., "usdc")
   * @param walletAddress - Wallet address to check
   */
  getTokenBalance(asset: string, walletAddress: Address): Promise<Decimal>;

  /**
   * Approve the contract to spend tokens for an asset
   * @param asset - Asset symbol (e.g., "usdc")
   * @param amount - Amount to approve
   * @returns Transaction hash
   */
  approve(asset: string, amount: Decimal): Promise<Hex>;

  // ========= Getters - ChannelsHub =========

  /**
   * Get the node's balance for a specific token
   * @param token - Token address
   * @param subId - Subaccount ID
   */
  getNodeBalance(token: Address, subId: number): Promise<Decimal>;

  /**
   * Get all open channel IDs for a user
   * @param user - User wallet address
   */
  getOpenChannels(user: Address): Promise<string[]>;

  /**
   * Get home channel data
   * @param homeChannelId - Home channel ID
   */
  getHomeChannelData(homeChannelId: string): Promise<HomeChannelDataResponse>;

  /**
   * Get escrow deposit data
   * @param escrowChannelId - Escrow channel ID
   */
  getEscrowDepositData(escrowChannelId: string): Promise<EscrowDepositDataResponse>;

  /**
   * Get escrow withdrawal data
   * @param escrowChannelId - Escrow channel ID
   */
  getEscrowWithdrawalData(escrowChannelId: string): Promise<EscrowWithdrawalDataResponse>;

  // ========= IVault functions =========

  /**
   * Deposit tokens into custody
   * @param node - Node address
   * @param token - Token address
   * @param amount - Amount to deposit
   * @returns Transaction hash
   */
  deposit(node: Address, token: Address, amount: Decimal): Promise<Hex>;

  /**
   * Withdraw tokens from custody
   * @param node - Node address
   * @param token - Token address
   * @param amount - Amount to withdraw
   * @returns Transaction hash
   */
  withdraw(node: Address, token: Address, amount: Decimal): Promise<Hex>;

  // ========= Channel lifecycle =========

  /**
   * Create a new channel
   * @param def - Channel definition
   * @param initCCS - Initial channel state
   * @returns Transaction hash
   */
  create(def: ChannelDefinition, initCCS: State): Promise<Hex>;

  /**
   * Migrate channel to this blockchain
   * @param def - Channel definition
   * @param candidate - Candidate state
   * @returns Transaction hash
   */
  migrateChannelHere(def: ChannelDefinition, candidate: State): Promise<Hex>;

  /**
   * Checkpoint a channel state on-chain
   * @param candidate - Candidate state
   * @returns Transaction hash
   */
  checkpoint(candidate: State): Promise<Hex>;

  /**
   * Challenge a channel state
   * @param candidate - Candidate state
   * @param challengerSig - Challenger signature
   * @param challengerIdx - Challenger participant index
   * @returns Transaction hash
   */
  challenge(candidate: State, challengerSig: Uint8Array, challengerIdx: number): Promise<Hex>;

  /**
   * Close a channel
   * @param candidate - Candidate state
   * @returns Transaction hash
   */
  close(candidate: State): Promise<Hex>;

  // ========= Escrow deposit =========

  /**
   * Initiate escrow deposit
   * @param def - Channel definition
   * @param initCCS - Initial channel state
   * @returns Transaction hash
   */
  initiateEscrowDeposit(def: ChannelDefinition, initCCS: State): Promise<Hex>;

  /**
   * Challenge escrow deposit
   * @param candidate - Candidate state
   * @param challengerSig - Challenger signature
   * @param challengerIdx - Challenger participant index
   * @returns Transaction hash
   */
  challengeEscrowDeposit(
    candidate: State,
    challengerSig: Uint8Array,
    challengerIdx: number,
  ): Promise<Hex>;

  /**
   * Finalize escrow deposit
   * @param candidate - Candidate state
   * @returns Transaction hash
   */
  finalizeEscrowDeposit(candidate: State): Promise<Hex>;

  // ========= Escrow withdrawal =========

  /**
   * Initiate escrow withdrawal
   * @param def - Channel definition
   * @param initCCS - Initial channel state
   * @returns Transaction hash
   */
  initiateEscrowWithdrawal(def: ChannelDefinition, initCCS: State): Promise<Hex>;

  /**
   * Challenge escrow withdrawal
   * @param candidate - Candidate state
   * @param challengerSig - Challenger signature
   * @param challengerIdx - Challenger participant index
   * @returns Transaction hash
   */
  challengeEscrowWithdrawal(
    candidate: State,
    challengerSig: Uint8Array,
    challengerIdx: number,
  ): Promise<Hex>;

  /**
   * Finalize escrow withdrawal
   * @param candidate - Candidate state
   * @returns Transaction hash
   */
  finalizeEscrowWithdrawal(candidate: State): Promise<Hex>;
}

// ============================================================================
// StateAdvancer Interface
// ============================================================================

/**
 * StateAdvancer validates state transitions
 */
export interface StateAdvancer {
  /**
   * Validate that proposedState is a valid advancement from currentState
   * @param currentState - Current state
   * @param proposedState - Proposed next state
   * @throws Error if advancement is invalid
   */
  validateAdvancement(currentState: State, proposedState: State): Promise<void>;
}

// ============================================================================
// StatePacker Interface
// ============================================================================

/**
 * StatePacker serializes channel states for on-chain submission
 */
export interface StatePacker {
  /**
   * Pack a state into bytes for on-chain submission
   * @param state - State to pack
   * @returns Packed bytes
   */
  packState(state: State): Promise<`0x${string}`>;
}

// ============================================================================
// AssetStore Interface
// ============================================================================

/**
 * AssetStore provides asset and token metadata
 */
export interface AssetStore {
  /**
   * Get asset decimals
   * @param asset - Asset symbol
   * @returns Number of decimals (uint8)
   */
  getAssetDecimals(asset: string): Promise<number>;

  /**
   * Get token decimals for a specific blockchain
   * @param blockchainId - Blockchain ID (uint64)
   * @param tokenAddress - Token contract address
   * @returns Number of decimals (uint8)
   */
  getTokenDecimals(blockchainId: bigint, tokenAddress: Address): Promise<number>;
}

// ============================================================================
// BlockchainEventHandler Interface
// ============================================================================

/**
 * BlockchainEventHandler handles channel lifecycle events from the blockchain
 */
export interface BlockchainEventHandler {
  /**
   * Handle home channel created event
   */
  handleHomeChannelCreated(event: HomeChannelCreatedEvent): Promise<void>;

  /**
   * Handle home channel migrated event
   */
  handleHomeChannelMigrated(event: HomeChannelMigratedEvent): Promise<void>;

  /**
   * Handle home channel checkpointed event
   */
  handleHomeChannelCheckpointed(event: HomeChannelCheckpointedEvent): Promise<void>;

  /**
   * Handle home channel challenged event
   */
  handleHomeChannelChallenged(event: HomeChannelChallengedEvent): Promise<void>;

  /**
   * Handle home channel closed event
   */
  handleHomeChannelClosed(event: HomeChannelClosedEvent): Promise<void>;

  /**
   * Handle escrow deposit initiated event
   */
  handleEscrowDepositInitiated(event: EscrowDepositInitiatedEvent): Promise<void>;

  /**
   * Handle escrow deposit challenged event
   */
  handleEscrowDepositChallenged(event: EscrowDepositChallengedEvent): Promise<void>;

  /**
   * Handle escrow deposit finalized event
   */
  handleEscrowDepositFinalized(event: EscrowDepositFinalizedEvent): Promise<void>;

  /**
   * Handle escrow withdrawal initiated event
   */
  handleEscrowWithdrawalInitiated(event: EscrowWithdrawalInitiatedEvent): Promise<void>;

  /**
   * Handle escrow withdrawal challenged event
   */
  handleEscrowWithdrawalChallenged(event: EscrowWithdrawalChallengedEvent): Promise<void>;

  /**
   * Handle escrow withdrawal finalized event
   */
  handleEscrowWithdrawalFinalized(event: EscrowWithdrawalFinalizedEvent): Promise<void>;
}

File: pkg/core/interface.go
package core

import (
	"context"

	"github.com/shopspring/decimal"
)

// ========= Client Interface =========

// Client defines the interface for interacting with the ChannelsHub smart contract
// TODO: add context to all methods
type BlockchainClient interface {
	// Getters - Token Balance & Approval
	GetTokenBalance(asset string, walletAddress string) (decimal.Decimal, error)
	Approve(asset string, amount decimal.Decimal) (string, error)

	// Getters - ChannelsHub
	GetNodeBalance(token string) (decimal.Decimal, error)
	GetOpenChannels(user string) ([]string, error)
	GetHomeChannelData(homeChannelID string) (HomeChannelDataResponse, error)
	GetEscrowDepositData(escrowChannelID string) (EscrowDepositDataResponse, error)
	GetEscrowWithdrawalData(escrowChannelID string) (EscrowWithdrawalDataResponse, error)

	// Node vault functions
	Deposit(token string, amount decimal.Decimal) (string, error)
	Withdraw(to, token string, amount decimal.Decimal) (string, error)

	// Node lifecycle
	EnsureSigValidatorRegistered(validatorID uint8, validatorAddress string, checkOnly bool) error

	// Channel lifecycle
	Create(def ChannelDefinition, initCCS State) (string, error)
	MigrateChannelHere(def ChannelDefinition, candidate State) (string, error)
	Checkpoint(candidate State) (string, error)
	Challenge(candidate State, challengerSig []byte, challengerIdx ChannelParticipant) (string, error)
	Close(candidate State) (string, error)

	// Escrow deposit
	InitiateEscrowDeposit(def ChannelDefinition, initCCS State) (string, error)
	ChallengeEscrowDeposit(candidate State, challengerSig []byte, challengerIdx ChannelParticipant) (string, error)
	FinalizeEscrowDeposit(candidate State) (string, error)

	// Escrow withdrawal
	InitiateEscrowWithdrawal(def ChannelDefinition, initCCS State) (string, error)
	ChallengeEscrowWithdrawal(candidate State, challengerSig []byte, challengerIdx ChannelParticipant) (string, error)
	FinalizeEscrowWithdrawal(candidate State) (string, error)
}

// ========= TransitionValidator Interface =========

// StateAdvancer applies state transitions
type StateAdvancer interface {
	ValidateAdvancement(currentState, proposedState State) error
}

// ========= StatePacker Interface =========

// StatePacker serializes channel states
type StatePacker interface {
	PackState(state State) ([]byte, error)
}

// ========= AssetStore Interface =========

type AssetStore interface {
	// GetAssetDecimals checks if an asset exists and returns its decimals in YN
	GetAssetDecimals(asset string) (uint8, error)

	// GetTokenDecimals returns the decimals for a token on a specific blockchain
	GetTokenDecimals(blockchainID uint64, tokenAddress string) (uint8, error)
}

// ReadOnlyChannelHub is a read-only view of the on-chain ChannelHub contract,
// used by event handlers to converge the Node row with chain after a guard
// drops an event. Each reactor binds a ReadOnlyChannelHub for its own chain
// and threads it into the handler methods that need an authoritative on-chain
// snapshot; no global multi-chain dispatcher is required.
type ReadOnlyChannelHub interface {
	// FetchChannel reads the authoritative on-chain channel snapshot for channelID
	// and returns a RefreshedChannel ready to overwrite the Node's local row. The
	// snapshot reflects on-chain state at RPC-read time, not event-emit time.
	FetchChannel(ctx context.Context, channelID string) (*RefreshedChannel, error)
}

// ChannelHubEventHandler defines the off-chain reactions to ChannelHub
// blockchain events. Only the three home-channel guard-drop handlers
// (HandleHomeChannelChallenged, HandleHomeChannelCheckpointed,
// HandleHomeChannelClosed) take a ReadOnlyChannelHub: they are the entrypoints
// where a version-regression guard may drop an event whose outer transaction
// has nonetheless committed state on chain, and the on-chain refresh is
// required to converge the Node row with chain. Other handlers do not need
// the hub and so do not accept the parameter, keeping the interface narrow.
type ChannelHubEventHandler interface {
	HandleNodeBalanceUpdated(context.Context, ChannelHubEventHandlerStore, *NodeBalanceUpdatedEvent) error
	HandleHomeChannelCreated(context.Context, ChannelHubEventHandlerStore, *HomeChannelCreatedEvent) error
	HandleHomeChannelMigrated(context.Context, ChannelHubEventHandlerStore, *HomeChannelMigratedEvent) error
	HandleHomeChannelCheckpointed(ctx context.Context, tx ChannelHubEventHandlerStore, hub ReadOnlyChannelHub, event *HomeChannelCheckpointedEvent) error
	HandleHomeChannelChallenged(ctx context.Context, tx ChannelHubEventHandlerStore, hub ReadOnlyChannelHub, event *HomeChannelChallengedEvent) error
	HandleHomeChannelClosed(ctx context.Context, tx ChannelHubEventHandlerStore, hub ReadOnlyChannelHub, event *HomeChannelClosedEvent) error
	HandleEscrowDepositInitiated(context.Context, ChannelHubEventHandlerStore, *EscrowDepositInitiatedEvent) error
	HandleEscrowDepositChallenged(context.Context, ChannelHubEventHandlerStore, *EscrowDepositChallengedEvent) error
	HandleEscrowDepositFinalized(context.Context, ChannelHubEventHandlerStore, *EscrowDepositFinalizedEvent) error
	HandleEscrowDepositsPurged(context.Context, ChannelHubEventHandlerStore, *EscrowDepositsPurgedEvent) error
	HandleEscrowWithdrawalInitiated(context.Context, ChannelHubEventHandlerStore, *EscrowWithdrawalInitiatedEvent) error
	HandleEscrowWithdrawalChallenged(context.Context, ChannelHubEventHandlerStore, *EscrowWithdrawalChallengedEvent) error
	HandleEscrowWithdrawalFinalized(context.Context, ChannelHubEventHandlerStore, *EscrowWithdrawalFinalizedEvent) error
}

type ChannelHubEventHandlerStore interface {
	// GetLastStateByChannelID retrieves the most recent state for a given channel.
	// If signed is true, only returns states with both user and node signatures.
	// Returns nil if no matching state exists.
	GetLastStateByChannelID(channelID string, signed bool) (*State, error)

	// GetLastUserState retrieves the most recent state for a user's asset across all
	// channels and detached chain entries (HomeChannelID nil). Returns nil if no
	// matching state exists. If signed is true, only fully co-signed states are returned.
	GetLastUserState(wallet, asset string, signed bool) (*State, error)

	// GetStateByChannelIDAndVersion retrieves a specific state version for a channel.
	// Returns nil if the state with the specified version does not exist.
	GetStateByChannelIDAndVersion(channelID string, version uint64) (*State, error)

	// UpdateChannel persists changes to a channel's metadata (status, version, etc).
	// The channel must already exist in the database.
	UpdateChannel(channel Channel) error

	// GetChannelByID retrieves a channel by its unique identifier.
	// Returns nil if the channel does not exist.
	GetChannelByID(channelID string) (*Channel, error)

	// ScheduleCheckpoint schedules a checkpoint operation for a home channel state.
	// This queues the state to be submitted on-chain to update the channel's on-chain state.
	ScheduleCheckpoint(stateID string, chainID uint64) error

	// ScheduleChallenge schedules a challengeChannel(...) submission on the channel's home
	// blockchain using the provided state and a node-produced challenger signature.
	ScheduleChallenge(stateID string, chainID uint64) error

	// ScheduleInitiateEscrowDeposit schedules an initiate for an escrow deposit operation.
	// This queues the state to be submitted on-chain to finalize an escrow deposit.
	ScheduleInitiateEscrowDeposit(stateID string, chainID uint64) error

	// ScheduleFinalizeEscrowDeposit schedules a finalize for an escrow deposit operation.
	// This queues the state to be submitted on-chain to finalize an escrow deposit.
	ScheduleFinalizeEscrowDeposit(stateID string, chainID uint64) error

	// ScheduleFinalizeEscrowWithdrawal schedules a checkpoint for an escrow withdrawal operation.
	// This queues the state to be submitted on-chain to finalize an escrow withdrawal.
	ScheduleFinalizeEscrowWithdrawal(stateID string, chainID uint64) error

	// SetNodeBalance upserts the on-chain liquidity for a given blockchain and asset.
	SetNodeBalance(blockchainID uint64, asset string, value decimal.Decimal) error

	// RefreshUserEnforcedBalance recomputes the locked balance from the user's open home channel on-chain state.
	RefreshUserEnforcedBalance(wallet, asset string) error

	// LockUserState acquires SELECT ... FOR UPDATE on the user's balance row so the
	// caller's transaction serializes against concurrent RPC paths that already lock
	// the same row before issuing receiver states. Postgres-only; SQLite is a no-op
	// in tests.
	LockUserState(wallet, asset string) (decimal.Decimal, error)

	// LockUserStateForHomeChannel locks the balance row of the user owning channelID. On
	// postgres it derives the lock key from the channel in SQL and returns the channel read
	// under that lock; on non-postgres (sqlite in tests) the snapshot is taken before the lock
	// for test compatibility. Event handlers must use this instead of a GetChannelByID +
	// LockUserState pair, which reads channel status before the lock and races a concurrent
	// submit_state finalization. Returns nil if the channel is absent.
	LockUserStateForHomeChannel(channelID string) (*Channel, error)

	// UpdateStateSigsIfMissing backfills the user and/or node signatures for a stored state
	// when the corresponding column is currently NULL. Used to repair the local record after
	// an on-chain event proves the state was enforced. Either signature may be empty to skip
	// that side; existing values are never overwritten and the call is idempotent on event replay.
	UpdateStateSigsIfMissing(channelID string, version uint64, userSig, nodeSig string) error

	// HasSignedFinalize reports whether a node-signed Finalize state exists for the given
	// home channel. Used to detect the post-Finalize lifecycle when the channel status
	// has been temporarily overwritten by an on-chain challenge.
	HasSignedFinalize(channelID string) (bool, error)

	// SumNetTransitionAmountAfterVersion returns the net effect on the user's
	// home-channel balance of transitions stored against channelID strictly above
	// minVersion. Receiver credits (TransferReceive, Release) contribute positively;
	// sender debits (TransferSend, Commit) contribute negatively. Other transition
	// kinds are excluded. Used to compute the ChallengeRescue amount when a
	// challenged channel is closed.
	SumNetTransitionAmountAfterVersion(channelID string, minVersion uint64) (decimal.Decimal, error)

	// StoreUserState persists a user state row. Used by the event handler to record a
	// ChallengeRescue squash state derived from a closed challenged channel.
	StoreUserState(state State, applicationID string) error

	// RecordTransaction creates a transaction row linking state transitions. Used by the
	// event handler to record the ChallengeRescue transaction associated with the squash.
	RecordTransaction(tx Transaction, applicationID string) error
}
